lib/messageReceiver.ts

```diff
@@ -163,29 +164,99 @@ export class MessageReceiver extends LinkEntity {
     this.maxConcurrentCalls = options.maxConcurrentCalls != undefined ?
       options.maxConcurrentCalls : 1;
     this.autoComplete = !!options.autoComplete;
-    this.receiveMode = options.receiveMode || ReceiveMode.peekLock;
+    this.maxAutoRenewDurationInSeconds = options.maxAutoRenewDurationInSeconds != undefined
+      ? options.maxAutoRenewDurationInSeconds
+      : 300;
+    this.autoRenewLock = this.maxAutoRenewDurationInSeconds > 0 && this.receiveMode === ReceiveMode.peekLock;
     this._onAmqpMessage = async (context: EventContext) => {
       const connectionId = this._context.namespace.connectionId;
       const bMessage: ServiceBusMessage = new ServiceBusMessage(this._context, context.message!,
         context.delivery!);
+      let timer: NodeJS.Timer | undefined = undefined;
+      let continueExecution: boolean = true;
+      if (this.autoRenewLock) {
+        // - We need to renew locks before they expire by looking at bMessage.lockedUntilUtc.
+        // - This autorenewal needs to happen **NO MORE** than maxAutoRenewDurationInSeconds
+        // - We should be able to clear the renewal timer when the user's message handler
+        // is done (whether it succeeds or fails).
+        log.receiver("[%s] message with id '%s' is locked until %s.",
+          connectionId, bMessage.messageId, bMessage.lockedUntilUtc!.toString());
+        const totalAutoLockRenewDuration = Date.now() + (this.maxAutoRenewDurationInSeconds * 1000);
+        log.receiver("[%s] Total autolockrenew duration for message with id '%s' is: ",
+          connectionId, bMessage.messageId, new Date(totalAutoLockRenewDuration).toString());
+        const autoRenewLockTask = (): void => {
+          if (Date.now() < totalAutoLockRenewDuration && continueExecution) {
+            const amount = calculateRenewAfterDuration(bMessage.lockedUntilUtc!);
+            log.receiver("[%s] Sleeping for %d milliseconds while renewing the lock for message " +
+              "with id '%s' is: ", connectionId, amount, bMessage.messageId);
+            timer = setTimeout(async () => {
+              try {
+                log.receiver("[%s] Attempting to renew the lock for message with id '%s'.",
+                  connectionId, bMessage.messageId);
+                await this._context.managementClient!.renewLock(bMessage);
+                log.receiver("[%s] Successfully renewed the lock for message with id '%s'.",
+                  connectionId, bMessage.messageId);
+              } catch (err) {
+                log.error("[%s] An error occured while auto renewing the message lock '%s' for " +
+                  "message with id '%s': %O.", connectionId, bMessage.lockToken,
+                  bMessage.messageId, err);
+              }
+              log.receiver("[%s] Calling the autorenewlock task again for message with id '%s'.",
+                connectionId, bMessage.messageId);
+              autoRenewLockTask();
```

Comment:
Why run this in the case of a failed lock renewal? If you're renewing at 75-80% of the total lock duration and then fail and wait another 75-80%, your lock will be long expired by the time things come back around again. You might consider building in some faster retries for specific types of transient failures, but if those fail, then the auto-renewal seems pretty much dead in the water for that message. I'd move the log and requeue code into the `try` branch of this function.

Another thing to be aware of in the future: you can renew locks for multiple messages in a single call. It's more complicated to manage coalescing the renewals, but you'll reduce the chattiness between the client and server, which may be worth it for high volume scenarios. I wouldn't worry about it now, but it's a good optimization to be aware of should you need it down the line.